design/private-repos.md

```diff
@@ -0,0 +1,72 @@
+# Screwdriver Access to Private Repositories
+
+## Context
+Provide a way for users to create and execute a pipeline for a private GitHub repository.
+
+This documentation covers the design proposal for that new feature.
+
+## Status
+
+- Oct 10, 2018: Design document
+
+## The Problem
+
+The current authentication methods do not provide access to private repositories.  This includes
+GitHub API calls as well as git clones via SSH.
+
+The OAuth token provided today for GitHub API calls is limited in scope to only public repositories.  If this scope were to 
+be expanded, then all users (whether they need this expanded scope or not) would be granting Screwdriver
+a security scope larger than required.  The screwdriver app should only be granted scope
+that is required.
+There are several stages of a Screwdriver pipeline in which this expanded scope is required
+when dealing with private repositories.  This document will outline each stage and how it will be addressed.
+
+Additionally, the Screwdriver app only has access to clone public repositories with the ssh key
+currently being used by Screwdriver.
+
+## Background
+
+Reference [Issue 1079][issue1079]
+
+## Proposal
+Any user intending to use Screwdriver with private repositories must perform the following prerequisites:
+- create a Personal Access Token within the Developer Settings of their Github user profile
+and include that token into their Screwdriver user profile.
+	- Github Personal Access Token is created at: https://github.com/settings/tokens/new
+	- Screwdriver user profile is edited at:  /user-settings
+- create/obtain an ssh keypair 
+	- store the public key within the Deploy Keys section of the private repository settings 
+	- associate the private key with the Screwdriver pipeline associated with this private repository
+
+Private Repository Pipeline Creation:
+1 A GitHub Personal Access Token that allows control of private repositories must be associated to the user profile within Screwdriver.
+1 User indicates on the Create form that the repository is a private repository which expands the web form to ask for additional information.
+1 User adds the private ssh key associated with the private repository to the Pipeline Create webform.
+
+The private ssh key will be stored as a Secret on the pipeline at pipeline creation time.  This Secret will be named [TBD]. 
+The existence of the secret with name [TBD] will essentially be a flag that indicates that this is a
+private repository requiring special handling.
+- When Screwdriver attempts to perform API calls for a private repository pipeline, it will use this flag to know to use the 
+GitHub Personal Access Token associated with the user logged in.
```

Comment:
Can we replace this access check with something else ? Rather then current check against a repository is it possible to replace this check with say isUserAdmin() at a GIT org level ? or even skip it if deploy keys are available.

This can be straight forward if it's a users private repo Eg: `jithine/private-repo` we can just check if logged in user is same as the SCM org.
But tricky when it's a teams organization `team/private`. Can we explore if there is any hook which GIT offers which ties permission check at an organization level ?

Alternatively if user does provide a deploy key, we could skip all checks which ties back to SCM and let it pass or fail at code checkout.

Idea is to see if we can get away without doing Personal Access Token.